### ISSUE TITLE
Interactive use and copying graphs

### ISSUE THREAD

**Issue by wangyuchen** [mdlincoln/clipr#42]
First, thank you for this awesome package! 

Before I discovered this package recently, I have been maintaining a package that has similar functionalities within my company for 3 years. It is not as thorough as `clipr` and it only works on Windows. However, there are some features in that package that are really handy. I'd love to know what you think about them and if you'd consider adding them to `clipr`. I'd be happy to make some pull requests. 

1. Use `.Last.value` as the default value for copying. 

    - It's great for interactive use. When an object is printed in the console, just type `clipr::write_clip()` in console and the object will be in the clipboard, instead of typing `clipr::write_clip(.Last.value)` (I found it hard to type `.Last.value` right).
    - I found that what users want to copy out is usually end results that they want to show others. In the R script, this is not necessarily an object that they want to save and reuse. And it's hard to come up with object names in the first place. In the current setting, the object need to be assigned and passed to `content`.
    - This is a feature that can be added to `write_clip()` directly or have a wrapper just for interactive use.
 
2. Now `write_clip()` returns either `rendered_content` or `content` based on `return_new`. As discussed [here](https://adv-r.hadley.nz/functions.html#invisible) in Advanced R, is it possible for `write_clip()` to always return the `content` invisibly? 
    - For a function that triggers a side effect, it's nice to be able to predict what it returns. 
    - Another reason is that the calls can be chained. See below for an example. Right now, the second call with `col.names = FALSE` won't work. The second call will take the rendered characters from the first call (with column names) and copy it. It will require `head(mtcars)` to be saved and passed explicitly, or `head(mtcars)` to be run and copied again. 

```
> head(mtcars)
                   mpg cyl disp  hp drat    wt  qsec vs am gear carb
Mazda RX4         21.0   6  160 110 3.90 2.620 16.46  0  1    4    4
Mazda RX4 Wag     21.0   6  160 110 3.90 2.875 17.02  0  1    4    4
Datsun 710        22.8   4  108  93 3.85 2.320 18.61  1  1    4    1
Hornet 4 Drive    21.4   6  258 110 3.08 3.215 19.44  1  0    3    1
Hornet Sportabout 18.7   8  360 175 3.15 3.440 17.02  0  0    3    2
Valiant           18.1   6  225 105 2.76 3.460 20.22  1  0    3    1
> clipr::write_clip(.Last.value)
> clipr::write_clip(.Last.value, col.names = FALSE)
```

If it returns `content` by default, and combined with `.Last.value` as default input, it can be used like this:

```
> head(mtcars)
                   mpg cyl disp  hp drat    wt  qsec vs am gear carb
Mazda RX4         21.0   6  160 110 3.90 2.620 16.46  0  1    4    4
Mazda RX4 Wag     21.0   6  160 110 3.90 2.875 17.02  0  1    4    4
Datsun 710        22.8   4  108  93 3.85 2.320 18.61  1  1    4    1
Hornet 4 Drive    21.4   6  258 110 3.08 3.215 19.44  1  0    3    1
Hornet Sportabout 18.7   8  360 175 3.15 3.440 17.02  0  0    3    2
Valiant           18.1   6  225 105 2.76 3.460 20.22  1  0    3    1
> clipr::write_clip()  # paste it out and check results
> clipr::write_clip(col.names = FALSE)  # not satisfy with results, call again with additional settings
```

3. Copying plots? Which I think `rstudioapi` already have the ability to do but it's not programmatic.
4. There may be other concerns about this, but providing `tibble` objects when reading tables? So when users paste in a long table, they won't see it taken over the whole screen as long as `tibble` is loaded..

Right now I have wrapper functions built on top of `clipr` for all these features (except for 3). Let me know if you'd consider having these as part of `clipr`. Thanks. 

Yuchen

**Comment by mdlincoln** [comment]
Thanks so much for your detailed message and suggestions @wangyuchen! My thoughts:

1. I like this a lot - I know I'd end up using it a lot based on my own patterns of `clipr` usage. Instead of modifying the default behavior of `write_clip()`, though, I'd probably prefer to export a new wrapper function like `write_last_clip()`.
2. Both `write_clip(..., return_new = TRUE)` and `write_clip(..., return_new = FALSE)` both return using `invisible()` already, so are you asking to change the default value of `return_new` to `FALSE`?
3. A nice idea, though I hesitate to dive into this since it would need to be multiplatform in order to make it into clipr. Though I do see https://unix.stackexchange.com/questions/145131/copy-image-from-clipboard-to-file ...
4. Hm, might consider this as long as there's an optional argument to return a vanilla `data.frame` - while I personally like everything that `tibbles` have to offer, the fact that they subtly modify `[` subsetting behavior could affect anyone who has been using clipr and expects it to return a non-tibble data.frame.

**Comment by wangyuchen** [comment]
1: Thanks! Adding a separate wrapper makes sense. I know that once it's been published even the smallest modification can have unexpected consequences.  

2 & 4: Same as 1, changing the current default behavior might have consequences, even if you leave a backdoor option for it. 

For `[` subsetting, I think `data.table` modified it most... Depending on if you have `data.table` loaded or not, a script writing for `data.table` may or may not work. I'm under the impression that `tibble` avoided this problem but I'm not 100% sure (I load `dplyr` everyday).  

I think this is all related to a more general question: whether a function's output value should rely on its arguments (e.g. `return_new = TRUE`, `data.frame = TRUE`). In interactive use, where `return_new = TRUE` is set explicitly, it's easy to predict what the function will return. But when used programmatically, i.e. when `return_new` is relying on another object or expression, the behavior of the function becomes somewhat unpredictable. 

What do you think of this as general guidelines:
1. adding more options to the current functions provided they don't change their current behavior.
2. providing user-facing wrapper functions that require less typing and are predictable. If users want to do something the higher-level wrapper functions doesn't do, it will require some more typing (longer function names, more arguments to set) based on the lower-level functions.


I have a working solution on clipping plots on Windows. I'll look into OSX and Linux options and let you know what I find. Would you accept pull requests for wrapper functions? I'd love to contribute if it's okay with you. I'd have to develop it for my internal package anyway..

**Comment by mdlincoln** [comment]
Sure, go ahead and add separate PRs for the wrappers so I can take a look!

**Comment by mdlincoln** [comment]
I've merged most of these requests with [0.6.0](https://github.com/mdlincoln/clipr/tree/v0.6.0). Thanks for your ideas!
